ignite/metrics/epoch_metric.py

```diff
@@ -68,21 +68,20 @@ def update(self, output: Sequence[torch.Tensor]) -> None:
         if y.ndimension() == 2 and y.shape[1] == 1:
             y = y.squeeze(dim=-1)
 
-        y_pred = y_pred.to(self._predictions)
-        y = y.to(self._targets)
-
-        self._predictions = torch.cat([self._predictions, y_pred], dim=0)
-        self._targets = torch.cat([self._targets, y], dim=0)
+        self._predictions.append(y_pred.clone())
```

Comment:
At first, we need to pass to CPU for both `y_pred` and `y`, otherwise we can overflow GPU's memory.
I'm also now think about whether we need to call `y_pred.detach()` to clearly stop the graph etc...

Reply:
> EDIT: it seems like CI is not happy with the changes.

From the error msg, it still take "self._predictions" as Tensor. That puzzled me.

Reply:
I don’t catch why it’s O(N^2) and why the implementation with append is better than cat. I would think that in both cases, memory should be allocated to receive the concat of existing array and new data. To me m, it’s more or less the same.

I must miss something but don’t know what.

Reply:
@sdesrozis: The original algorithm is O(N^2) because:
torch.cat([A, B]) returns a new tensor instead of copying B to the end of A. Thus, it needs to allocate memory and iterate the contents of A every time. It is the same as the following loop:
```
for i in range(N):
    for j in range(i):
        # read data in A[j]
```
I noticed this because the running time of EpochMetric is multiple fold longer than the forward pass of the same data (even with dummy metric that only return length). And, it is worse when I reduce the batch_size.

For the new proposed approach, I believe torch.cat([...]) only allocates one time, and iterate the content of each tensor once.

Reply:
The advantage of tensors over list is to be in GPU memory.

Reply:
Yes, it depends a lot on the implementation of torch.cat(). Let me find someone with the knowledge to review. In addition, we actually can benchmark this with big data.

Reply:
Related discussions on PyTorch forum : 
- https://discuss.pytorch.org/t/appending-in-pytorch/39313
- https://discuss.pytorch.org/t/appending-to-a-tensor/2665/3

Torch.cat is a good solution. To improve performance, we could preallocate as suggested. But I agree, we need to bench.

Reply:
Ok I misunderstood the implementation !! 

You do a list of tensors and use torch.cat only one time. Sorry for my reading...